tools/online_editor/src/index.ts

```diff
@@ -40,28 +48,209 @@ function create_demo_menu(editor: EditorWidget): Menu {
+    label: "Copy Permalink to Clipboard",
+    iconClass: "fa da-share",
+    mnemonic: 1,
+    execute: () => {
+      const params = new URLSearchParams();
+      params.set("snippet", editor.current_editor_content);
+      const this_url = new URL(window.location.toString());
+      this_url.search = params.toString();
+      const share_url = this_url.toString();
+
+      navigator.clipboard.writeText(share_url);
+    },
+  });
+
   menu.addItem({ command: "slint:copy_permalink" });
 
   return menu;
 }
 
-function main() {
-  const preview = new PreviewWidget();
-  const editor = new EditorWidget();
-  const properties = new PropertiesWidget();
-  const welcome = new WelcomeWidget();
+function widget_pseudo_url(id: string): string {
+  return "::widget:://" + id;
+}
 
-  editor.onRenderRequest = (style, source, url, fetcher) => {
-    return preview.render(style, source, url, fetcher);
-  };
-  editor.onNewPropertyData = (binding_text_provider, p) => {
-    properties.set_properties(binding_text_provider, p);
-  };
+function id_from_pseudo_url(url: string): string {
+  const id = url.substring(14);
+  console.assert(url == widget_pseudo_url(id));
+  return id;
+}
+
+function create_view_menu(dock_widgets: DockWidgets): Menu {
+  const dock = dock_widgets.dock;
+
+  const menu = new Menu({ commands });
+  menu.title.label = "Views";
+
+  commands.addCommand("slint:save-dock-layout", {
+    label: "Save Dock Layout",
+    caption: "Save the current dock layout",
+    execute: () => {
+      const layout = dock.saveLayout();
+      const widgets = Array.from(dock.widgets());
+
+      const objMap: Map<object, string> = new Map();
+      for (const w of widgets) {
+        objMap.set(w, widget_pseudo_url(w.title.label));
+      }
+      objMap.set(dock, "::widget:://DockPanel");
+      if (dock.layout != null) {
+        objMap.set(dock.layout, "::object:://DockLayout");
+      }
+      if (dock.parent != null) {
+        objMap.set(dock.parent, "::widget:://parent");
+      }
+      try {
+        const layout_str = JSON.stringify(layout, (_, value) => {
+          const result = objMap.get(value);
+          if (result === undefined) {
+            return value;
+          }
+          return result;
+        });
+        localStorage.setItem("layout", layout_str);
+      } catch (e) {
+        console.log("Failed to save dock layout!", e);
+      }
+    },
+  });
+
+  commands.addCommand("slint:reset-dock-layout", {
+    label: "Reset to Default Layout",
+    caption: "Reset to default layout",
+    execute: () => {
+      localStorage.removeItem("layout");
+    },
+  });
+
+  commands.addCommand("slint:restore-dock-layout", {
+    label: "Restore Dock Layout",
+    caption: "Restore the stored Dock layout",
+    execute: () => {
+      const layout_str = localStorage.getItem("layout");
+      if (layout_str != null && layout_str != "") {
+        const idMap: Map<string, Widget | Layout | null | undefined> =
```

Comment:
I think that this map instance can't really return `null` or `undefined`, can it?

Reply:
I need to check that: There were definitely null and undefined in that map at some point, but I might have removed the need for those in the meantime.

Reply:
The `null` is needed as not all possible widgets might exist at a given moment. But `undefined` can indeed go now.